Feature/remove redialing

i-hate-nicknames opened the pull request:
Did you run `make format && make check`?

Fixes #830
Fixes #844
Fixes #843

 Changes:	
- Remove transport redialing
- Remove transport statuses as a concept
- Add persistent transports configuration (transports that are repeatedly redialed when broken, and are set up at visor startup)

How to test this PR:
- Run a couple of visors and establish transports between them manually, delete those transports manually, crash visors deliberately.
- Use transport setup to add skywire transports and public visors and check that the connections are properly get set up and are not reconnected.
- Add persistent transports to configuration and check that it gets established on startup and reestablished when the target visor goes offline and back

A known issue is that sudph transports do not get closed automatically, when other end of connection closes. This could be an issue with kcp-go library configuration, or it could be the case that the library doesn't support it.
For this case, [I added an automatic connection close logic](https://github.com/skycoin/skywire/pull/842/files#diff-9a822d96295268985a33b3260cfe31fae865ab929e735feb6e2427e2b1687a57R330), built on `net.Conn` deadline logic. Basically, the idea is to update deadline every time the connection is used.
For this to work properly, we need to add support for some kind of keep-alive messages to the transport layer. 
Currently, it is disabled (deadline is 0 and ignored by conn). 

Shutdown logic is also fixed in this PR, since I thought it would help with sudph issues.

i-hate-nicknames committed:
Remove redialing of transport

i-hate-nicknames committed:
Fix panics when transport is closed

i-hate-nicknames committed:
Refactor managed transport Serve

i-hate-nicknames committed:
Force sorted edge order in transport entry

i-hate-nicknames committed:
Remove up/down transport status

i-hate-nicknames committed:
Remove closed transports in transport manager

i-hate-nicknames committed:
Move raw address to network.Conn

i-hate-nicknames committed:
Simplify dial logic

i-hate-nicknames committed:
Delete entry instead of updating status

i-hate-nicknames committed:
Fix visor cli transport print

i-hate-nicknames committed:
Add persistent transports todo

i-hate-nicknames committed:
Add heartbeat loop

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -66,16 +56,8 @@ type ManagedTransport struct {
 
-	dc  DiscoveryClient
-	ls  LogStore
-	ebc *appevent.Broadcaster
```
We still need this structure though? Adding new transports at runtime is required to trigger events to be broadcasted to VPN for example. Am I missing somehting?

i-hate-nicknames replied:
I checked and they were not used anywhere else except for some tcp events. I moved those events to transport manager itself to avoid passing those around, but perhaps they should be kept here.

jdknives reviewed (commented):
Looks good. Well done.

jdknives commented:
Please also remove the `Char` file. Not related to this PR but it should not be in the repo.

i-hate-nicknames committed:
Merge branch 'develop' into feature/remove-redialing

i-hate-nicknames committed:
Send heartbeat request in transport

i-hate-nicknames committed:
Remove transport status

i-hate-nicknames committed:
Fix linter errors

i-hate-nicknames committed:
Remove transport entry with status

i-hate-nicknames committed:
Fix visor cli transports

i-hate-nicknames committed:
Add persistent transports config

i-hate-nicknames committed:
Make lint and format

i-hate-nicknames committed:
Fix deadlocks in transport layer

i-hate-nicknames committed:
Relax locking when saving transport

Previously on transport save the whole manager would be blocked
until the whole process is finished

Now, saveTransport method only locks in critical sections, and sometimes
uses lighter locks (RLock). This allows multiple saveTransport methods
to run simultaneously

i-hate-nicknames committed:
Add transport connection deadline

i-hate-nicknames committed:
Fix transport manager shutdown logic

i-hate-nicknames committed:
Simplify transport manager closing

i-hate-nicknames committed:
Merge branch 'develop' into feature/remove-redialing

i-hate-nicknames committed:
Remove isPublic transport entry field

mrpalide commented on cmd/skywire-cli/commands/visor/transport_discovery.go:
```diff
@@ -58,13 +58,13 @@ var discTpCmd = &cobra.Command{
-		_, err := fmt.Fprintf(w, "%s\t%s\t%t\t%d\t%t\t%s\t%s\t%t\t%t\n",
-			e.Entry.ID, e.Entry.Type, e.Entry.Public, e.Registered, e.IsUp, e.Entry.Edges[0], e.Entry.Edges[1], e.Statuses[0], e.Statuses[1])
+		_, err := fmt.Fprintf(w, "%s\t%s\t%s\t%s\n",
+			e.ID, e.Type, e.Edges[0], e.Edges[1])
```
The `public` column is extra and should be removed from header.

ersonp commented on pkg/transport/managed_transport.go:
```diff
@@ -532,25 +394,18 @@ func (mt *ManagedTransport) WritePacket(ctx context.Context, packet routing.Pack
-
-			return fmt.Errorf("failed to redial underlying connection: %w", err)
-		}
+		return fmt.Errorf("write packet: cannot write to conn, conn is not set up")
```
We should use `fmt.Error` instead.

i-hate-nicknames replied:
There is only Errorf in fmt

ersonp replied:
Ya, my bad. I confused it with log.

ersonp commented on pkg/transport/manager.go:
```diff
@@ -330,43 +384,25 @@ func (tm *Manager) saveTransport(ctx context.Context, remote cipher.PubKey, init
-	}()
+	go mTp.Serve(tm.readCh)
+	tm.mx.Lock()
+	tm.Logger.Debugf("Locked in saveTransport")
```
We should use `Debug` here.

ersonp commented on pkg/transport/manager.go:
```diff
@@ -330,43 +384,25 @@ func (tm *Manager) saveTransport(ctx context.Context, remote cipher.PubKey, init
+	tm.Logger.Debugf("Locked in saveTransport")
 	tm.tps[tpID] = mTp
+	tm.mx.Unlock()
+	tm.Logger.Debugf("Unlocked in saveTransport")
```
We should use `Debug` here.

ersonp commented on pkg/transport/manager.go:
```diff
@@ -390,7 +427,9 @@ func (tm *Manager) STCPRRemoteAddrs() []string {
 	tm.mx.Lock()
+	tm.Logger.Debugf("Locked in DeleteTransport")
 	defer tm.mx.Unlock()
+	defer tm.Logger.Debugf("Unlocked in DeleteTransport")
```
We should use `Debug` here too.

ersonp commented on pkg/transport/managed_transport.go:
```diff
@@ -137,83 +103,77 @@ func (mt *ManagedTransport) Serve(readCh chan<- routing.Packet) {
+		case <-ticker.C:
+			err := mt.dc.HeartBeat(context.Background(), mt.Entry.ID)
+			if err != nil {
+				log.Warnf("Failed to send heartbeat")
```
We should use `fmt.Warn` here.

i-hate-nicknames committed:
Fix NPE for transport handshake

i-hate-nicknames committed:
Code review fixes

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -13,16 +13,19 @@ import (
-const logWriteInterval = time.Second * 3
+const (
+	logWriteInterval  = time.Second * 3
+	heartbeatInterval = time.Second * 45
```
We probably should increase that interval a lot. Something like 10 minutes would be fine. The reason we need that heartbeat is that it is possible two visors are crashing and unable to deregister a transport. Transports entries could then remain in the discovery without ever being cleaned up. This scenario should not happen very frequently. On the other hand, the more transport we have, the more this heartbeat will be straining the deployment. This is only needed to make sure we are not keeping entries indefinitely in the discovery.

jdknives commented on pkg/transport/handshake.go:
```diff
@@ -99,7 +96,7 @@ func (hs SettlementHS) Do(ctx context.Context, dc DiscoveryClient, conn network.
-		entry := makeEntryFromTpConn(conn, true)
+		entry := makeEntryFromTpConn(conn)
 
 		// TODO(evanlinjin): Probably not needed as this is called in mTp already. Need to double check.
```
This is not yet part of the PR but should be. There is a similar defer call a bit further down.

ersonp replied:
We do not have `status` anymore no I don't think this is needed anymore.

jdknives replied:
@ersonp I agree. Please delete.

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -232,27 +192,47 @@ func (mt *ManagedTransport) isServing() bool {
-func (mt *ManagedTransport) disconnect() {
-	mt.once.Do(func() { close(mt.done) })
-	_ = mt.updateStatus(false, 1) //nolint:errcheck
+// close underlying connection and update entry status in transport discovery
```
This should state that we remove from discovery instead of update the status.

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -396,10 +326,19 @@ func (mt *ManagedTransport) getConn() network.Conn {
 }
 
+// updates underlying connection inactive deadline
+func (mt *ManagedTransport) updateConnDeadline() {
```
I would tear the deadlines out as we discussed. We currently are not setting them anywhere and adding them back if necessary should be easy enough. So essentially this is dead code. As discussed in DMs, this is mostly necessary for `sudph` as it cant be closed properly atm so we should look into that specific transport first.

jdknives reviewed (changes requested):
We should rename the `Conn` Interface to `transport interface as that seems to be what it is judging from our current naming conventions (i.e. the structure encapsulating it being called `managed transport`) and from spec.

jdknives commented:
@ersonp as @i-hate-nicknames is out, I would like you to take over the rest of the comments as they are not too broad in scope or deep in effect. After changes are made, make sure to give a good bit of testing to the overall PR and we should get this merged.

ersonp committed:
Update heartbeat interval

ersonp committed:
Update comment

ersonp committed:
Remove updateConnDeadline

ersonp committed:
Rename Conn interface to Transport

ersonp committed:
Remove EntryWithStatus

ersonp committed:
Add Deprecated Public field to Entry struct

ersonp committed:
Make public true

ersonp committed:
Update and add todo

mrpalide reviewed (approved):
Everything looks good, and testing between this PR and master work without problem.

ersonp committed:
Remove old todo

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -50,8 +50,8 @@ type ManagedTransportConfig struct {
-// There is a single underlying connection between two edges.
-// Initial dialing can be requested by either edge of the connection.
+// There is a single underlying transport connection between two edges.
+// Initial dialing can be requested by either edge of the transport connection.
```
`There is a single underlying transport between two edges`. I would not chain together transport and connection in this way. A transport is a connection simply.

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -208,7 +208,7 @@ func (mt *ManagedTransport) IsClosed() bool {
 }
 
-// close underlying connection and remove the entry from transport discovery
+// close underlying transport connection and remove the entry from transport discovery
```
Same note about usage of transport and connection as above.

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -221,34 +221,34 @@ func (mt *ManagedTransport) close() {
-func (mt *ManagedTransport) Accept(ctx context.Context, conn network.Conn) error {
-	mt.connMx.Lock()
-	defer mt.connMx.Unlock()
+// Accept accepts a new underlying transport connection.
```
Update comment.

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -257,24 +257,24 @@ func (mt *ManagedTransport) Accept(ctx context.Context, conn network.Conn) error
 
-	mt.log.Debug("Setting underlying connection...")
-	return mt.setConn(conn)
+	mt.log.Debug("Setting underlying transport connection...")
```
Same applies here.

jdknives commented on pkg/transport/managed_transport.go:
```diff
@@ -287,20 +287,20 @@ func (mt *ManagedTransport) DialAsync(ctx context.Context, errCh chan error) {
 
 func (mt *ManagedTransport) dial(ctx context.Context) error {
-	conn, err := mt.client.Dial(ctx, mt.rPK, skyenv.DmsgTransportPort)
+	transportType, err := mt.client.Dial(ctx, mt.rPK, skyenv.DmsgTransportPort)
```
That seems to return a transport itself and not the type of the transport.

jdknives reviewed (commented):
Please check for and remove all occurences of `transport connection` and rename it to plain `transport`. Having both is redundant and therefore confusing. Also a few small comments otherwise.

ersonp committed:
Remove public field from Entry struct

ersonp committed:
Rename PersistentRemote to PersistentTransports

ersonp committed:
Requested changes

ersonp committed:
Add api for persistent_transports

jdknives reviewed (commented):
API looks good! Now that I see it, I am not sure if Senyoret actually needs the `GetPersistentTransports` or whether it would be more convenient to get it in the Summary call. @Senyoret1 let us know.

Senyoret1 commented:
If persistent and non-persistent transports are going to be shown just as the transports are being shown at this moment in the UI, having all the data in the summary would be the best option, as no changes in the UI would be needed. If you want the UI to do something special with persistent transports, then I would need more info to know what would have to be done and what will be nedded

ersonp commented:
We will need to show the list of persistent transports set up by the user as well as a way for the user to add or remove them. Is there anything more that needs to be done @jdknives

ersonp committed:
Add persistent_transports to summary

ersonp committed:
Remove GetPersistentTransports api

ersonp committed:
Merge branch 'develop' into feature/remove-redialing
